hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/PartialDisconnectionHandler.java

```diff
@@ -0,0 +1,149 @@
+
+import com.hazelcast.cluster.impl.MemberImpl;
+import com.hazelcast.internal.util.graph.BronKerboschCliqueFinder;
+import com.hazelcast.internal.util.graph.Graph;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static java.util.Collections.emptySet;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+class PartialDisconnectionHandler {
+    private final long detectionIntervalMs;
+    private final long algorithmTimeoutMs;
+    private Map<MemberImpl, Set<MemberImpl>> disconnections = new HashMap<>();
+    private long lastUpdated;
+
+    PartialDisconnectionHandler(long detectionIntervalMs, long algorithmTimeoutMs) {
+        this.detectionIntervalMs = detectionIntervalMs;
+        this.algorithmTimeoutMs = algorithmTimeoutMs;
+    }
+
+    boolean update(MemberImpl member, long timestamp, Collection<MemberImpl> disconnectedMembers) {
+        if (timestamp < lastUpdated) {
+            return false;
+        }
+
+        Set<MemberImpl> currentDisconnectedMembers =  disconnections.get(member);
+        if (currentDisconnectedMembers == null) {
+            if (disconnectedMembers.isEmpty()) {
+                return false;
+            }
+
+            currentDisconnectedMembers = new HashSet<>();
+            disconnections.put(member, currentDisconnectedMembers);
+        }
+
+        boolean updated = false;
+
+        for (MemberImpl disconnectedMember : disconnectedMembers) {
+            if (currentDisconnectedMembers.add(disconnectedMember)
+                    && !disconnections.getOrDefault(disconnectedMember, emptySet()).contains(member)) {
+                lastUpdated = timestamp;
+                updated = true;
+            }
+        }
+
+        if (currentDisconnectedMembers.retainAll(disconnectedMembers)) {
+            lastUpdated = timestamp;
+            updated = true;
+        }
+
+        if (currentDisconnectedMembers.isEmpty()) {
+            disconnections.remove(member);
+        }
+
+        return updated;
+    }
+
+    boolean shouldResolvePartialDisconnections(long timestamp) {
+        return !disconnections.isEmpty() && timestamp - lastUpdated >= detectionIntervalMs;
+    }
+
+    Collection<MemberImpl> resolve(Map<MemberImpl, Set<MemberImpl>> disconnections) {
+        Set<MemberImpl> members = new HashSet<>();
+        disconnections.forEach((k, v) -> {
+            members.add(k);
+            members.addAll(v);
+        });
+
+        Graph<MemberImpl> connectivityGraph = buildConnectionGraph(members, disconnections);
+        BronKerboschCliqueFinder<MemberImpl> cliqueFinder = createCliqueFinder(connectivityGraph);
+        Collection<Set<MemberImpl>> maxCliques = cliqueFinder.computeMaxCliques();
+
+        if (cliqueFinder.isTimeLimitReached()) {
+            throw new IllegalStateException("Partial disconnection resolution algorithm timed out! disconnectivity map: "
```

Comment:
Is this really an illegal state: `IllegalStateException` ?  maybe something like `TimeoutException` would be better..